File: src/gonet/client/lua/packet.lua
require("crc")
local pb = require "pb"
local protoc = require "protoc"
m_PacketCreateMap = {}
m_PacketMap = {}


-- 加载pb
assert(pb.loadfile "./../../message/pb/message.pb")
assert(pb.loadfile "./../../message/pb/game.pb")
assert(pb.loadfile "./../../message/pb/client.pb")

--创建包头
function BuildPacketHead(id, destservertype)
    return{
         Stx = 0x27,
         DestServerType = destservertype,
         Ckx=0x72,
         Id = id
    }
end

--包处理回调函数
function RegisterPacket(packetName, func)
    name = string.lower(packetName)
    id = CRC32.hash(name)
    m_PacketCreateMap[id] = "message." .. packetName
    m_PacketMap[id] = func
end

--处理包函数
function HandlePacket(dat)
    id = bytes_to_int(string.sub(dat, 0, 4))
    buff = string.sub(dat, 5)
    packetName = m_PacketCreateMap[id]
    if packetName ~= nil then
        local packet = pb.decode(packetName, buff)
        m_PacketMap[id](packet)
    end
end

--发送包函数
function SendPacket(name, packet
    id = crc32.Gen(string.lower(name))
    packetName = "message." .. name
    if packetName ~= nil then
        local bytes = pb.encode(packetName, packet)
        bytes = int_to_bytes(id) .. bytes .. TCP_END
        CLIENTSOCKET:Send(bytes)
    end
end

function bytes_to_int(str,endian,signed) -- use length of string to determine 8,16,32,64 bits
    local t={str:byte(1,-1)}
    if endian=="big" then --reverse bytes
        local tt={}
        for k=1,#t do
            tt[#t-k+1]=t[k]
        end
        t=tt
    end
    local n=0
    for k=1,#t do
        n=n+t[k]*2^((k-1)*8)
    end
    if signed then
        n = (n > 2^(#t-1) -1) and (n - 2^#t) or n -- if last bit set, negative.
    end
    return n
end

function int_to_bytes(num,endian,signed)
    if num<0 and not signed then num=-num print"warning, dropping sign from number converting to unsigned" end
    local res={}
    local n = math.ceil(select(2,math.frexp(num))/8) -- number of bytes to be used.
    if signed and num < 0 then
        num = num + 2^n
    end
    for k=n,1,-1 do -- 256 = 2^8 bits per char.
        local mul=2^(8*(k-1))
        res[k]=math.floor(num/mul)
        num=num-res[k]*mul
    end
    assert(num==0)
    if endian == "big" then
        local t={}
        for k=1,n do
            t[k]=res[n-k+1]
        end
        res=t
    end
    return string.char(table.unpack(res))
end

 -- lua table data
 local data = {
    PacketHead = {
         Stx = 0x27,
         DestServerType = 2,
         Ckx=0x72,
         Id = 0
    },
    Sender = 1000,
    SenderName = "test"
 }

 --local bytes = assert(pb.encode("message.W_C_ChatMessage", data))
 --print(pb.type("message.W_C_ChatMessage"))
 -- encode lua table data into binary format in lua string and return
 --local data2 = assert(pb.decode("message.W_C_ChatMessage", bytes))
 --print(require "serpent".block(data2))

File: src/gonet/client/js/Packet.js
//IPacket.js
var CRC32 = require('crc-32'); 
var Network = require("./network");
var BitStream = require("./BitStream");
var messagepb= require("./pb/message");
var clientpb= require("./pb/client");
STX = 0X27;
CKX = 0x72;

SERVICE_NONE          = 0;
SERVICE_CLIENT        = 1;
SERVICE_GATESERVER    = 2;
SERVICE_ACCOUNTSERVER = 3;
SERVICE_WORLDSERVER   = 4;
SERVICE_MONITORSERVER = 5;

Uint8Array.prototype.indexOfMulti = function(searchElements, fromIndex) {
    fromIndex = fromIndex || 0;

    var index = Array.prototype.indexOf.call(this, searchElements[0], fromIndex);
    if(searchElements.length === 1 || index === -1) {
        // Not found or no other elements to check
        return index;
    }

    for(var i = index, j = 0; j < searchElements.length && i < this.length; i++, j++) {
        if(this[i] !== searchElements[j]) {
            return this.indexOfMulti(searchElements, index + 1);
        }
    }

    return (i === index + searchElements.length) ? index : -1;
};

var Packet = (function(){
    //var Packet = {
        //整形转换成字节
        IntToBytes =  function(v){
            var b = new Uint8Array(4);
            b[0] = v
            b[1] = v >> 8
            b[2] = v >> 16
            b[3] = v >> 24
            return b
        },

        //字节转换成整形
        BytesToInt =   function(b){
            return b[0] | b[1]<<8 | b[2]<<16 | b[3]<<24;
        },


        //创建包头
        BuildPacketHead =  function(Id, DestServerType=0){
            var pHead =  new messagepb.message.Ipacket.create();
            pHead.Ckx = CKX;
            pHead.Stx = STX;
            pHead.Id = Id;
            pHead.DestServerType = DestServerType;
            return pHead
        },

        /*Uint8ArrayToString = function(fileData){
              var dataString = "";
              for (var i = 0; i < fileData.length; i++) {
                dataString += String.fromCharCode(fileData[i]);
              }
             
              return dataString
        },

        stringToUint8Array = function(str){
              var arr = [];
              for (var i = 0, j = str.length; i < j; ++i) {
                arr.push(str.charCodeAt(i));
              }
             
              var tmpUint8Array = new Uint8Array(arr);
              return tmpUint8Array
        },*/

        //包处理回调函数
        RegisterPacket =  function(packetName, func){
            var name = packetName.toLowerCase();
            var id = CRC32.str(name);
            if(m_callbacks[id] != null)
            {
                console.log("消息[%s]重复注册!!!!!!!!!!!!!",packetName);
                return;
            }

            m_callbacks[id] = func;
        },

        //--处理包函数
        HandlePacket  =  function(dat){
            var id = BytesToInt(dat.slice(0,4));
            var buf = dat.slice(4);
            var packetcreator = m_packets[id];
            console.log(id, packetcreator, m_packets, packet, m_callbacks);
            var packet = packetcreator().decode(buf);
            m_callbacks[id](packet);
            return true;
        },


        //--发送包函数
        SendPacket = function(name, dat){
            var id = CRC32.str(name.toLowerCase());
            var packetName = "message." + name;
            if(packetName != null)
            {
                var packetcreator = m_packets[id];
                var crc =IntToBytes(id);
                var packetcreator = m_packets[id];
                var packet = packetcreator().encode(dat).finish();
                var buff = new Uint8Array(crc.length + packet.length + TCP_END_LENGTH);
                buff.set(crc, 0);
                buff.set(packet, crc.length);
                buff.set(TCP_END, crc.length+packet.length);
                //var buff = Uint8ArrayToString(IntToBytes(id)) + dat + TCP_END;
                Network.getInstance().send(buff);
            }
            return false;
        },

        //解析包
        ReceivePacket =  function(dat){
            seekToTcpEnd = function(dat){
                var nCurLen = dat.indexOfMulti(TCP_END);
                if(nCurLen != -1){
                    return nCurLen + TCP_END_LENGTH;
                }
                return 0;
            }

            var buff = new Uint8Array(m_pInBuffer.length + dat.length);
            buff.set(m_pInBuffer, 0);
            buff.set(dat, m_pInBuffer.length);
            m_pInBuffer = new Uint8Array(0);
            var nCurSize = 0
            ParsePacekt = function(){
                var nPacketSize = 0
                var buff1 = buff.slice(nCurSize)
                var nBufferSize = buff1.length;
                nPacketSize = seekToTcpEnd(buff1)
                //console.log(nPacketSize, nBufferSize, TCP_END_LENGTH)
                if (nPacketSize != 0){
                    if(nBufferSize == nPacketSize){ //完整包
                        //print(string.sub(buff1, 0, nPacketSize - TCP_END_LENGTH))
                        HandlePacket(buff1.slice(0, nPacketSize - TCP_END_LENGTH))
                        nCurSize =  nCurSize + nPacketSize
                    }
                    else if (nBufferSize > nPacketSize){
                        //print(string.sub(buff1, 0, nPacketSize - TCP_END_LENGTH))
                        HandlePacket(buff1.slice(0, nPacketSize - TCP_END_LENGTH))
                        nCurSize =  nCurSize + nPacketSize
                        ParsePacekt();
                    }
                    else if(nBufferSize < 128 * 1024){
                        self.m_pInBuffer = buff.slice(nCurSize)
                    }
                    else{
                        console.log("超出最大包限制，丢弃该包")
                    }
                }
            },

            ParsePacekt();
        };
    //};
})();
//var TCP_END  = "💞♡";
var TCP_END = new Uint8Array(7);
TCP_END.set( [240,159,146,158,226,153,161],0);
var TCP_END_LENGTH  = TCP_END.length;
var m_pInBuffer = new Uint8Array();
var m_callbacks = new Array();
var m_packets = new Array();

//--protobuf-- 自动化解析
function RegisterPacketCreator(name, func){
    var id = CRC32.str(name.toLowerCase());
    m_packets[id] = func;
};

RegisterPacketCreator("W_C_SelectPlayerResponse", function(){
    return clientpb.message.W_C_SelectPlayerResponse;
})
RegisterPacketCreator("C_A_LoginRequest", function(){
    return clientpb.message.C_A_LoginRequest;
})

Network.getInstance().host = "ws://192.168.84.62:31700/ws";
Network.getInstance().initNetwork();
module.exports = Packet;